### PR TITLE
handle chat completion request with rag

### PR DESCRIPTION
**Pull request by guangrui-cloudnatix**

### CODE REVIEW

**Comment by guangrui-cloudnatix** [comment]
to be tested e2e.

**Review by kkaneda — APPROVED** [review]
LG!

File: api/v1/inference_server.proto [hunk 1]
```diff
@@ -44,6 +44,15 @@ message CreateChatCompletionRequest {
     Function function = 2;
   }
 
+  message ToolChoice {
+    string choice = 1;
```
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply api/v1/inference_server.proto:48]: Not sure what's the best way to implement `tool_choice` in proto, maybe using [Value](https://protobuf.dev/reference/protobuf/google.protobuf/#google.protobuf.Value)?

File: server/internal/rag/rag.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"fmt"
+	"log"
+
+	v1 "github.com/llm-operator/inference-manager/api/v1"
+	vsv1 "github.com/llm-operator/vector-store-manager/api/v1"
+	"google.golang.org/grpc"
+)
+
+const (
+	prompt string = "Answer the users QUESTION using the DOCUMENT text below. Keep your answer ground in the facts of the DOCUMENT. If the DOCUMENT doesn’t contain the facts to answer the QUESTION return {NONE}"
```
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply server/internal/rag/rag.go:14]: refer to https://community.openai.com/t/prompt-engineering-for-rag/621495/3

File: server/internal/rag/rag.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			return nil, err
+		}
+		for _, doc := range searchResp.Documents {
+			msgs = append(msgs, &v1.CreateChatCompletionRequest_Message{
+				Content: doc,
+				Role:    "assistant",
+			})
+		}
+	}
+	if len(msgs) > 0 {
```
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply server/internal/rag/rag.go:82]: Refer to https://platform.openai.com/docs/guides/text-generation/chat-completions-api as an example

File: server/internal/server/completions.go [hunk 4]
```diff
@@ -143,3 +159,42 @@ func (s *S) CreateChatCompletion(
 		return
 	}
 }
+
+// handleTools uses tools to process and modify the request messages.
+// Refer to https://platform.openai.com/docs/guides/function-calling for the details of the tools and the tool choice.
```
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply server/internal/server/completions.go:164]: I don't have an example implementation for this. I just thought we can use `tools` for it.

File: server/internal/server/completions.go [hunk 5]
```diff
@@ -143,3 +159,42 @@ func (s *S) CreateChatCompletion(
...
+			}
+			if ragFunction.VectorStoreName == "" {
+				return fmt.Errorf("vector store name is required")
+			}
+			msgs, err := s.rewritter.ProcessMessages(ctx, ragFunction.VectorStoreName, req.Messages)
+			if err != nil {
+				return err
+			}
+			req.Messages = msgs
+			return nil
```
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply server/internal/server/completions.go:194]: actually, `continue` is better.

File: server/internal/rag/rag.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kenji-cloudnatix) [reply server/internal/rag/rag.go:14]: Thanks! Will take a look.

File: api/v1/inference_server.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (guangrui-cloudnatix) [reply api/v1/inference_server.proto:48]: how do we set `none`, `auto`, or `required` string value?

[the remark this answers is not in the document]
Comment (kkaneda) [reply api/v1/inference_server.proto:48]: Ah now I understand. This was to represent the string value of `tool_choice`. 

I think we can use https://protobuf.dev/programming-guides/proto3/#any , but having a TODO comment also looks good to me.

File: server/internal/rag/rag.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkaneda) [reply server/internal/rag/rag.go:14]: How about adding the above link as code comment?

**Comment by guangrui-cloudnatix** [comment]
```
curl http://localhost:8080/v1/chat/completions -d '{
  "model": "google-gemma-2b-it-q4",
  "messages": [{"role": "user", "content": "What is Denver Art Museum?"}],
  "tool_choice": {
    "choice": "auto",
    "type": "function",
    "function": {
      "name": "rag"
    }
  },
  "tools": [{
    "type": "function",
    "function": {
      "name": "rag",
      "parameters": "{\"vector_store_name\":\"gfu-test1\"}"
    }
  }]
}'
{"id":"chatcmpl-168","object":"chat.completion","created":1720562386,"model":"google-gemma-2b-it-q4","system_fingerprint":"fp_ollama","choices":[{"index":0,"message":{"role":"assistant","content":"The Denver Art Museum is a non-profit cultural institution in Denver, Colorado, United States. The museum features a comprehensive collection of modern and contemporary art, with a special focus on European and American art. The museum also has a rich history, having been founded in 1903."},"finish_reason":"stop"}],"usage":{"prompt_tokens":1490,"completion_tokens":60,"total_tokens":1550}}
```

```
curl http://localhost:8080/v1/chat/completions -d '{
  "model": "google-gemma-2b-it-q4",
  "messages": [{"role": "user", "content": "What is cloud?"}],
  "tool_choice": {
    "choice": "auto",
    "type": "function",
    "function": {
      "name": "rag"
    }
  },
  "tools": [{
    "type": "function",
    "function": {
      "name": "rag",
      "parameters": "{\"vector_store_name\":\"gfu-test1\"}"
    }
  }]
}'
{"id":"chatcmpl-105","object":"chat.completion","created":1720560891,"model":"google-gemma-2b-it-q4","system_fingerprint":"fp_ollama","choices":[{"index":0,"message":{"role":"assistant","content":"The context does not mention anything about cloud, so I cannot answer this question from the provided context."},"finish_reason":"stop"}],"usage":{"prompt_tokens":1465,"completion_tokens":21,"total_tokens":1486}}
```

**Pull request merged**